cli/cmd/upgrade.go

```diff
@@ -106,16 +106,30 @@ func (options *upgradeOptions) validateAndBuild(k kubernetes.Interface, flags *p
 	options.overrideConfigs(configs, map[string]string{})
 	configs.GetInstall().Flags = options.recordedFlags
 
-	values, err := options.buildValuesWithoutIdentity(configs)
-	if err != nil {
-		upgradeErrorf("Could not build install configuration: %s", err)
+	var identity *installIdentityValues
+	idctx := configs.GetGlobal().GetIdentityContext()
+	if idctx.GetTrustDomain() == "" || idctx.GetTrustAnchorsPem() == "" {
+		// If there wasn't an idctx, or if it doesn't specify the required fields, we
+		// must be upgrading from a version that didn't support identity, so generate it anew...
+		identity, err = options.installOptions.identityOptions.genValues()
+		if err != nil {
+			upgradeErrorf("Unable to generate issuer credentials.\nError: %s", err)
+		}
+		configs.GetGlobal().IdentityContext = identity.toIdentityContext()
+	} else {
+		identity, err = fetchIdentityValues(k, options.controllerReplicas, idctx)
+		if err != nil {
+			upgradeErrorf("Unable to fetch the existing issuer credentials from Kubernetes.\nError: %s", err)
+		}
 	}
 
-	identityValues, err := fetchIdentityValues(k, options.controllerReplicas, configs.GetGlobal().GetIdentityContext())
+	// Values have to be generated after any missing identity is generated,
+	// otherwise it will be missing from the generated configmap.
+	values, err := options.buildValuesWithoutIdentity(configs)
```

Comment:
Confirmed successful upgrade from `edge-19.3.2` to this branch.

Related, the `autoInjectContext` is populated as `{}` via `install` but `null` via `upgrade`:
```bash
$ bin/linkerd install --ignore-cluster --proxy-auto-inject | grep autoInjectContext
..."autoInjectContext":{}}

$ bin/linkerd upgrade --proxy-auto-inject | grep autoInjectContext
..."autoInjectContext":null}
```

I think if we did something like `configs := options.configs(identityValues.toIdentityContext())`, we'd get a correctly populated `autoInjectContext` field.

Reply:
i don't want to call `proxyInjectOptions.configs` here, since i want to use the "original" config object and just update it from options (rather than generate a new config object entirely).

https://github.com/linkerd/linkerd2/commit/ccbf846e01ae5dda33c46e14b34b9dae1fd599b9 overrides the `overrideConfigs` receiver for `upgradeOptions` so that it can override the config with the autoinject settings.

This used to work but regressed in https://github.com/linkerd/linkerd2/commit/31706e541758d754ab873d6f854a0b3028713346.